### DISCUSSION TITLE
`middleware` `provideDocumentSymbols` doesn't display children symbols

### DISCUSSION THREAD

**Discussion by ttax00** [microsoft/vscode-discussions#213] [Extension Development QnA] [answered]
Concerning piece of code. https://code.visualstudio.com/api/language-extensions/embedded-languages
Calling the command on `.html` virtual document. 
```ts
async provideDocumentSymbols(document, token, next) {
let result: undefined | DocumentSymbol[] = undefined;

result = await commands.executeCommand<DocumentSymbol[]>(
'vscode.executeDocumentSymbolProvider',
vdocUri(document),
);

return result;
}
```
this returns an array of `DocumentSymbol[]` which has `children: DocumentSymbol[]` properties. But vscode only display the top level symbols instead of everything. See screenshots here: https://github.com/TechTheAwesome/code-yew-server/issues/2

A quick fix right now is to manually unpack the childrens into a flat DocumentSymbol array. Also, on startup, `vscode.executeDocumentSymbolProvider` would return undefined. And requires a few calls before the correct parse would be returned.

**Expected Behavior**
VSCode should display all childrens.

**Comment by alefragnani** [comment, marked as answer]
Hi @TechTheAwesome ,

Yes, the `vscode.executeDocumentSymbolProvider` command indeed returns a hierarchical structure, instead of a flat one, and it seems, based on https://github.com/microsoft/vscode/issues/58793, this approach decision was made a long time ago. 

I don't see any open issue about it today so, if you really think a flat return should be available, I suggest you to open an issue in VS Code repo (https://github.com/microsoft/vscode/issues/new?assignees=&labels=&template=feature_request.md), asking for such change/update in the command.

Hope this helps

**Reply by ttax00** [reply]
Thanks for your response @alefragnani,

I did encounter the issue when I was searching about the problem. But I don't think requesting change is what I am looking for?
 
My problem is `vscode.executeDocumentSymbolProvider` provides a hierarchical array, but `async provideDocumentSymbols(document, token, next)` hook seems to expecting a flattened array. I'm taking HTML symbols, and return it into RUST language environment, and I was wondering if its a language configuration issue, or perhaps its a bug with  `provideDocumentSymbols()`? 

My code follows this example here: https://code.visualstudio.com/api/language-extensions/embedded-languages
What I get:
![](https://user-images.githubusercontent.com/10691398/199187583-fced92fd-74ee-4bec-938d-e126a7106785.png)
What I am expecting:
![image](https://user-images.githubusercontent.com/10691398/200101999-8a96c86e-8cdf-4a44-b7cf-1f7933abfe7c.png)

**Reply by alefragnani** [reply]
Based on that issue, I don't think it is a bug, but the _designed behavior_. That's why I suggested you to open a feature request, so the VS Code team could take a look and, if necessary, update the command to provide a flattened list.

I use the same API on my **Separators** extension and, at first sight, I was expecting to receive the flattened list, but I understood it to be an hierarchical structure. I though it would be for performance reasons, and simply used it as it is.

Maybe another extension developer (or someone from the VS Code team) has an alternative command/API that works with flattened list. Then, maybe I could update my Separators extension to use the same 😃

**Reply by ttax00** [reply]
Ahh so you do have a need to use the symbols in order to place Separators correctly in your extension. I'll notify if the vscode team got back to me, but in my use case it shouldn't have matter which format the symbols data is in, if that make sense. What I'm doing is just getting the output from command, and return it without change to vscode, and in that case, vscode should've treated the data correctly? It feels like even though the output is hierarchical, the input is flattened, so its make me unsure whats the best approach here.   😅
